public/js/main.js

```diff
@@ -133,14 +128,23 @@ async function initialize() {
  * inference engine.
  */
 async function main() {
-  await initialize();
+  try {
```

Comment:
I'm ok if we submit like this (don't want to hold you up) but generally I'm not a huge fan of large try-catch blocks. Note that my philosophy is derived from other languages, not super sure about js. 

* Generally it's good practice to do explicit error handling. try-catch is useful when the lower level libraries you're calling into are going to raise errors at various points that you don't want to deal with individually - but here, we're dealing with our own libraries, all of which does its own error handling, so we get to decide if the lower level library calls should signal its failure status by raising an error or through its return value (either is fine by me, though i'm biased by golang and c++ where it's done through return value) - but if we decide to go the error route (as we do currently) then the try-catch should only surround the initialize() call, no the other bits.

* We shouldn't tell the user to "Reload the app" unless we have encountered a retriable error - init should be able to signal .
* We shouldn't have any code that touches the DOM in main.js (we likely want to have a ui.error(message) that does the dom manipulation (and great idea of disabling the buttons))

Reply:
> I'm ok if we submit like this (don't want to hold you up) but generally I'm not a huge fan of large try-catch blocks. Note that my philosophy is derived from other languages, not super sure about js.

I'm also not a fan and was considering doing something like:
```
async function main() {
  try {
    await initialize();
  } catch (err) {
    document.querySelector('h2').innerHTML +=
      '<br><em style="color: red;">Please reload this page!</em>';
    console.warn('App failure:', err);
    return -1; //or any value that would enable an early exit
  }
 
  setStateUI(state);
  initButtonsUI(exportData);
  initVideoUI();
  const canvas = initCanvas();

  startEngine(canvas, state.video);
}
```

But ESLint was getting in the way due to the inconsistent rule and I didn't feel it was a good idea to just ignore it and felt the bigger `try ... catch` way was cleaner and kept the `void` return type of the function.

I guess doing it the C-way (i.e. using `return 0` at the end of the function would sort that out.


> We shouldn't tell the user to "Reload the app" unless we have encountered a retriable error - init should be able to signal .

True, although (for now) there are no ways (that I know of at least) to get the app to automatically reload once the webcam is freed.
We could always get it to reload its `location` which would refresh the page but if the webcam is _still_ being used elsewhere that wouldn't do anything useful.

> We shouldn't have any code that touches the DOM in main.js (we likely want to have a ui.error(message) that does the dom manipulation (and great idea of disabling the buttons))

Good call, I'll implement that and remove the DOM manipulation from `main`.